### PR TITLE
Swift: Specify preference for structs over classes

### PR DESCRIPTION
**Pull request by gfontenot**
Also note that if you _are_ going to use a `class`, you should default
to marking it as `final`.

### CODE REVIEW

**Comment by gfontenot** [comment]
I'm going to add this info to the commit message, but the reasoning behind this change:

`struct`s are immutable value objects. This means that the compiler is able to perform better optimizations since it knows that the object can't change. It also reduces the cognitive overhead around mutation and state passing. `struct`s are also incapable of using inheritance, which encourages the use of `protocol`s and composition instead, which we prefer as a design pattern.

If you're going to use a `class`, marking it as `final` will ensure that it can't be subclassed. This lets the compiler make some better optimizations since it knows the methods being added won't ever be overridden. It also encourages the use of composition over inheritance, which we prefer as a design pattern.

**Comment by gilesvangruisen** [comment]
:+1:

**Comment by dgalarza** [comment]
:+1:

File: style/swift/README.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ Swift
 
 * Keep up with the Objective-C style guide above. Will highlight differences
   here.
+* Prefer `struct`s over `class`es wherever possible
+* If you need to use a `class`, default to marking it as `final`
```
Comment (jakecraige) [inline style/swift/README.md]: The wording of this is sort of based off the previous rule, what about rewording it so it stands on it's own? Something like "Default to marking classes as `final`"?

**Comment by gfontenot** [comment]
ping @thoughtbot/ios

**Comment by seanpdoyle** [comment]
Is there a way to codify this guideline in a [SwiftLint](https://houndci.com/configuration#swift) rule?

File: style/swift/README.md (hunk 1 above)
Comment (kaishin) [inline style/swift/README.md]: Agreed.

**Comment by dazmuda** [comment]
Do you think we should also mention:
Prefer protocols over classes, if possible.

It's more like... if it's possible and makes sense but also it makes sense and is possible more often than one would think.

**Comment by tonyd256** [comment]
:+1: to this concept

**Comment by gfontenot** [comment]
There don't seem to be any objections to this. I'm going to go ahead and merge. @dazmuda I'll open a PR today about preferring protocols to inheritance, which is I believe what you were going for?

**Pull request merged**